src/Tools/dotnet-monitor/Egress/AzureBlob/AzureBlobEgressProvider.cs

```diff
@@ -129,6 +157,91 @@ private Uri GetAccountUri(AzureBlobEgressProviderOptions options, out string acc
+
+            queueUriBuilder.Query = null;
+            queueUriBuilder.QueueName = null;
+
+            accountName = queueUriBuilder.AccountName;
+
+            return queueUriBuilder.ToUri();
+        }
+
+        public async Task EgressMessageToQueue(string artifactName, AzureBlobEgressProviderOptions options, CancellationToken token)
+        {
+            try
+            {
+                QueueClient queueClient = await GetQueueClientAsync(options, token);
+
+                if (queueClient.Exists())
+                {
+                    string queueMessage = ConstructQueueMessage(artifactName, options);
+                    queueClient.SendMessage(queueMessage);
+                }
+            }
+            catch (AggregateException ex) when (ex.InnerException is RequestFailedException innerException)
+            {
+                throw CreateException(innerException);
+            }
+            catch (RequestFailedException ex)
+            {
+                throw CreateException(ex);
+            }
+        }
+
+        private string ConstructQueueMessage(string artifactName, AzureBlobEgressProviderOptions options)
+        {
+            using var stream = new MemoryStream();
+            using var writer = new Utf8JsonWriter(stream);
+
+            writer.WriteStartObject();
+
+            writer.WriteString("artifact", artifactName);
+            writer.WriteString("uri", options.AccountUri.ToString());
+
+            writer.WriteEndObject();
+            writer.Flush();
+
+            return Encoding.UTF8.GetString(stream.ToArray());
+        }
+
+        private async Task<QueueClient> GetQueueClientAsync(AzureBlobEgressProviderOptions options, CancellationToken token)
+        {
+            QueueServiceClient serviceClient;
+            if (!string.IsNullOrWhiteSpace(options.SharedAccessSignature))
+            {
+                var serviceUriBuilder = new UriBuilder(options.QueueAccountUri)
+                {
+                    Query = options.SharedAccessSignature
+                };
+
+                serviceClient = new QueueServiceClient(serviceUriBuilder.Uri);
+            }
+            else if (!string.IsNullOrEmpty(options.AccountKey))
+            {
+                // Remove Query in case SAS token was specified
+                Uri accountUri = GetQueueAccountUri(options, out string accountName);
+
+                StorageSharedKeyCredential credential = new StorageSharedKeyCredential(
+                    accountName,
+                    options.AccountKey);
+
+                serviceClient = new QueueServiceClient(accountUri, credential);
+            }
+            else
+            {
+                throw CreateException(Strings.ErrorMessage_EgressMissingSasOrKey);
+            }
+
+            QueueClient queueClient = serviceClient.GetQueueClient(options.QueueName);
+
+            await queueClient.CreateIfNotExistsAsync(cancellationToken: token);
```

Comment:
The SAS token may allow for putting items into the queue but not allow for creation of the queue itself. Please explore if there is a way to check that the queue exists with a limited SAS token and then fallback to trying to call CreateIfNotExistsAsync. We have the same issue with the blob storage portion already but let's not pile more onto it: #445

Reply:
Off the top of your head, any reason why we couldn't just use `queueClient.Exists()` here? Looking at documentation, it says in the Remarks section for `CreateIfNotExistsAsync`, "This API requires Create or Write permissions," but there's no disclaimer for `Exists()`. I figure we could do an `Exists` check, and then if it doesn't exist we fall back to `CreateIfNotExistsAsync`?

Reply:
That would probably work, but use `ExistsAsync` (I think that's a method). Add a comment as to why we do "if !ExistsAsync() { CreateIfNotExistsAsync() }" because it won't be obvious why we shouldn't just call `CreateIfNotExistsAsync`.